cpp/src/arrow/adapters/orc/jni/jni_wrapper.cpp

```diff
@@ -0,0 +1,243 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <arrow/adapters/orc/adapter.h>
+#include <arrow/array.h>
+#include <arrow/buffer.h>
+#include <arrow/io/api.h>
+#include <arrow/ipc/api.h>
+#include <arrow/util/logging.h>
+#include <cassert>
+#include <string>
+
+#include "org_apache_arrow_adapter_orc_OrcMemoryJniWrapper.h"
+#include "org_apache_arrow_adapter_orc_OrcReaderJniWrapper.h"
+#include "org_apache_arrow_adapter_orc_OrcStripeReaderJniWrapper.h"
+
+using ORCFileReader = arrow::adapters::orc::ORCFileReader;
+using RecordBatchReader = arrow::RecordBatchReader;
+
+void ThrowJavaIOException(JNIEnv* env, const std::exception& e) {
+  jclass ioExceptionClass = env->FindClass("java/io/IOException");
+  if (ioExceptionClass != nullptr) {
+    if (env->ThrowNew(ioExceptionClass, e.what())) {
+      // Failed to new IOException. This means another error has occurred in Java
+      // We just propagate this error to caller by doing nothing.
+      ARROW_LOG(ERROR) << "Error occurred when throwing IOException";
+    }
+  } else {
+    ARROW_LOG(ERROR) << "Error occurred when getting IOException class";
+  }
+}
+
+void ThrowJavaException(JNIEnv* env, const std::string& message) {
+  jclass exception = env->FindClass("java/lang/Exception");
+  if (exception != nullptr) {
+    env->ThrowNew(exception, message.c_str());
+  } else {
+    throw std::runtime_error("Can't find java/lang/Exception class");
```

Comment:
generally we try to avoid exceptions in the code base, does this propagate back up to java or does it essentially crash the program.  if it crashes consider using ARROW_LOG(FATAL)